File: mailblastr-cli/src/commands/automations.js
'use strict';

const { collect, parseJson, clean, withPagination, pagination } = require('../helpers');

function register({ group, leaf, act }) {
  const automations = group('automations', 'Manage automations (scoped to a sending domain)');

  act(
    leaf(automations, 'create', 'Create an automation')
      .requiredOption('--name <name>', 'automation name')
      .requiredOption('--domain <domain>', 'the sending domain this automation belongs to')
      .option(
        '--trigger <event>',
        "trigger event, e.g. 'contact.created', 'mailblastr:schedule' (requires --trigger-at/--trigger-timezone), or a custom event name",
      )
      .option('--trigger-at <iso>', "ISO 8601 instant the 'mailblastr:schedule' trigger fires (with --trigger-timezone)")
      .option('--trigger-timezone <tz>', "IANA timezone the schedule was picked in, e.g. 'America/New_York'")
      .option('--status <status>', "'enabled' | 'disabled' (default: disabled)"),
    ({ client, opts }) =>
      client.automations.create(
        clean({
          name: opts.name,
          domain: opts.domain,
          trigger: opts.trigger,
          trigger_config:
            opts.triggerAt || opts.triggerTimezone
              ? clean({ at: opts.triggerAt, timezone: opts.triggerTimezone })
              : undefined,
          status: opts.status,
        }),
      ),
  );

  act(leaf(automations, 'get <id>', 'Retrieve an automation'), ({ client, args: [id] }) =>
    client.automations.get(id),
  );

  act(withPagination(leaf(automations, 'list', 'List automations')), ({ client, opts }) =>
    client.automations.list(pagination(opts)),
  );

  act(
    leaf(automations, 'update <id>', 'Update an automation')
      .option('--name <name>', 'new name')
      .option('--status <status>', "'enabled' | 'disabled'")
      .option('--trigger-at <iso>', "update the 'mailblastr:schedule' trigger's ISO 8601 fire time (with --trigger-timezone)")
      .option('--trigger-timezone <tz>', "IANA timezone the schedule was picked in, e.g. 'America/New_York'"),
    ({ client, opts, args: [id] }) =>
      client.automations.update(
        id,
        clean({
          name: opts.name,
          status: opts.status,
          trigger_config:
            opts.triggerAt || opts.triggerTimezone
              ? clean({ at: opts.triggerAt, timezone: opts.triggerTimezone })
              : undefined,
        }),
      ),
  );

  act(
    leaf(automations, 'add-step <id>', 'Append a step to an automation')
      .requiredOption(
        '--type <type>',
        "step type: 'send_email' | 'wait_for_event' | 'delay' | 'condition' | 'split' | 'add_to_segment' | 'contact_update' | 'contact_delete'",
      )
      .option('--config <json>', "step config as JSON, e.g. '{\"template_id\":\"tmpl_x\"}'")
      .option('--key <key>', 'step key (for connections)'),
    ({ client, opts, args: [id] }) =>
      client.automations.addStep(
        id,
        clean({ type: opts.type, config: parseJson(opts.config, '--config'), key: opts.key }),
      ),
  );

  act(
    leaf(automations, 'update-step <id> <stepId>', 'Update a step of an automation')
      .option('--type <type>', 'new step type')
      .option('--config <json>', 'new step config as JSON (replaces the stored config)')
      .option('--key <key>', 'new step key (for connections)'),
    ({ client, opts, args: [id, stepId] }) =>
      client.automations.updateStep(
        id,
        stepId,
        clean({ type: opts.type, config: parseJson(opts.config, '--config'), key: opts.key }),
      ),
  );

  act(
    leaf(automations, 'delete-step <id> <stepId>', 'Delete a step from an automation'),
    ({ client, args: [id, stepId] }) => client.automations.deleteStep(id, stepId),
  );

  act(
    withPagination(leaf(automations, 'runs <id>', "List an automation's runs, newest first")).option(
      '--status <status>',
      "keep only runs in these statuses: 'running' | 'completed' | 'failed' | 'skipped' (repeatable or comma-separated); filtering happens before paging",
      collect,
    ),
    ({ client, opts, args: [id] }) =>
      client.automations.runs(id, clean({ ...pagination(opts), status: opts.status })),
  );

  act(
    leaf(automations, 'run <id> <runId>', 'Retrieve a single automation run'),
    ({ client, args: [id, runId] }) => client.automations.getRun(id, runId),
  );

  act(
    leaf(automations, 'stop <id>', 'Stop an automation (in-progress runs finish)'),
    ({ client, args: [id] }) => client.automations.stop(id),
  );

  act(
    leaf(automations, 'ai <id>', 'Author (or extend) the step graph from a natural-language prompt')
      .requiredOption('--prompt <text>', 'what the automation should do (max 2000 characters)')
      .option('--template-id <id>', 'a template the plan may send (repeatable or comma-separated; first 10 used)', collect)
      .option('--event <name>', 'an event name the plan may wait on (repeatable or comma-separated; first 10 used)', collect)
      .option('--attach-from <key>', 'extend an existing graph from this trigger key or step key instead of authoring a whole workflow')
      .option('--attach-type <type>', "connection type: 'default' (default) | 'condition_met' | 'condition_not_met' | 'event_received' | 'timeout'")
      .option('--attach-before <key>', 'insert before this existing step key'),
    ({ client, opts, args: [id] }) =>
      client.automations.createWithAi(
        id,
        clean({
          prompt: opts.prompt,
          template_ids: opts.templateId,
          events: opts.event,
          // Presence of `attach` is what switches the API from workflow mode to
          // append mode, so only build it when the user named a source step.
          attach: opts.attachFrom
            ? clean({ from: opts.attachFrom, type: opts.attachType, before: opts.attachBefore })
            : undefined,
        }),
      ),
  );

  act(leaf(automations, 'delete <id>', 'Delete an automation'), ({ client, args: [id] }) =>
    client.automations.remove(id),
  );
}

module.exports = { register };

File: mailblastr-dotnet/src/Mailblastr/Models/Automations.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mailblastr;

/// <summary>A step of an automation as returned by the API.</summary>
public class AutomationStep
{
    /// <summary>Absent on the synthesized trigger step.</summary>
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("key")]
    public string Key { get; set; } = null!;

    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    [JsonPropertyName("position")]
    public int? Position { get; set; }

    [JsonPropertyName("config")]
    public Dictionary<string, JsonElement>? Config { get; set; }
}

/// <summary>A typed edge between two step keys.</summary>
public class AutomationConnection
{
    [JsonPropertyName("from")]
    public string From { get; set; } = null!;

    [JsonPropertyName("to")]
    public string To { get; set; } = null!;

    [JsonPropertyName("type")]
    public string? Type { get; set; }
}

/// <summary>An inline step supplied on automation create.</summary>
public class AutomationStepInput
{
    /// <summary>Optional key referenced by connections.</summary>
    [JsonPropertyName("key")]
    public string? Key { get; set; }

    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    [JsonPropertyName("config")]
    public Dictionary<string, object?>? Config { get; set; }
}

/// <summary>Enrollment counts (retrieve only).</summary>
public class AutomationEnrollments
{
    [JsonPropertyName("active")]
    public int Active { get; set; }

    [JsonPropertyName("completed")]
    public int Completed { get; set; }
}

/// <summary>An automation (multi-step workflow triggered by an event).</summary>
public class Automation
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "automation";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("audience_id")]
    public string AudienceId { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("trigger")]
    public string? Trigger { get; set; }

    /// <summary>
    /// The sending domain this automation belongs to. Only EventSendAsync calls
    /// with the same domain trigger it. Null on pre-domain rows (treated as the
    /// account's single domain when exactly one exists).
    /// </summary>
    [JsonPropertyName("domain")]
    public string? Domain { get; set; }

    /// <summary><c>enabled</c> | <c>disabled</c>.</summary>
    [JsonPropertyName("status")]
    public string Status { get; set; } = null!;

    /// <summary>
    /// Schedule of the <c>mailblastr:schedule</c> trigger; null for every other
    /// trigger.
    /// </summary>
    [JsonPropertyName("trigger_config")]
    public AutomationTriggerConfig? TriggerConfig { get; set; }

    /// <summary>
    /// Key of the synthetic trigger step that <c>connections[].from</c> edges
    /// reference; defaults to the literal <c>trigger</c>.
    /// </summary>
    [JsonPropertyName("trigger_key")]
    public string? TriggerKey { get; set; }

    /// <summary>
    /// The step graph. Omitted entirely on LIST responses (GET /automations) —
    /// retrieve the automation for its steps.
    /// </summary>
    [JsonPropertyName("steps")]
    public List<AutomationStep>? Steps { get; set; }

    [JsonPropertyName("connections")]
    public List<AutomationConnection>? Connections { get; set; }

    /// <summary>Enrollment counts — included only on retrieve (GET /automations/:id).</summary>
    [JsonPropertyName("enrollments")]
    public AutomationEnrollments? Enrollments { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;

    [JsonPropertyName("updated_at")]
    public string? UpdatedAt { get; set; }
}

/// <summary>
/// Config for the <c>mailblastr:schedule</c> trigger: the automation fires ONCE
/// at <see cref="At"/>, enrolling every contact of its domain's pool. Required
/// with that trigger; not accepted on any other.
/// </summary>
public class AutomationTriggerConfig
{
    /// <summary>ISO 8601 instant the automation fires (future, at most 366 days ahead).</summary>
    [JsonPropertyName("at")]
    public string At { get; set; } = null!;

    /// <summary>IANA timezone the schedule was picked in (e.g. <c>America/New_York</c>).</summary>
    [JsonPropertyName("timezone")]
    public string Timezone { get; set; } = null!;
}

/// <summary>Payload for creating an automation (POST /automations). <see cref="Domain"/> is REQUIRED.</summary>
public class AutomationCreateOptions
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    /// <summary>
    /// REQUIRED. The sending domain this automation belongs to (e.g.
    /// <c>yourdomain.com</c> — one of your domains). Only EventSendAsync calls
    /// with the same domain trigger it.
    /// </summary>
    [JsonPropertyName("domain")]
    public string Domain { get; set; } = null!;

    /// <summary>
    /// The event that starts a run: <c>contact.created</c>, the built-in
    /// scheduled trigger <c>mailblastr:schedule</c> (requires
    /// <see cref="TriggerConfig"/>), an engagement event (<c>email.opened</c>,
    /// <c>email.clicked</c>, <c>email.replied</c>, <c>email.bounced</c>,
    /// <c>email.delivered</c>), or any custom event name you send via
    /// EventSendAsync. Usually supplied as a Steps[0] trigger step instead.
    /// </summary>
    [JsonPropertyName("trigger")]
    public string? Trigger { get; set; }

    /// <summary>
    /// Schedule for the <c>mailblastr:schedule</c> trigger (<c>{at, timezone}</c>).
    /// Required with that trigger; not accepted on any other.
    /// </summary>
    [JsonPropertyName("trigger_config")]
    public AutomationTriggerConfig? TriggerConfig { get; set; }

    /// <summary>Initial status: <c>enabled</c> | <c>disabled</c> (default <c>disabled</c>).</summary>
    [JsonPropertyName("status")]
    public string? Status { get; set; }

    /// <summary>Optional inline step graph; each step may carry a Key for connections.</summary>
    [JsonPropertyName("steps")]
    public List<AutomationStepInput>? Steps { get; set; }

    /// <summary>Optional typed edges between step keys.</summary>
    [JsonPropertyName("connections")]
    public List<AutomationConnection>? Connections { get; set; }
}

/// <summary>Payload for updating an automation (PATCH /automations/:id).</summary>
public class AutomationUpdateOptions
{
    [JsonPropertyName("name")]
    public string? Name { get; set; }

    /// <summary><c>enabled</c> | <c>disabled</c>.</summary>
    [JsonPropertyName("status")]
    public string? Status { get; set; }

    /// <summary>Re-point the automation at another of your domains (disabled automations only).</summary>
    [JsonPropertyName("domain")]
    public string? Domain { get; set; }

    /// <summary>
    /// Change the triggering event (disabled automations only). The reserved
    /// <c>mailblastr:</c> prefix is rejected apart from <c>mailblastr:schedule</c>.
    /// </summary>
    [JsonPropertyName("trigger")]
    public string? Trigger { get; set; }

    /// <summary>
    /// Rename the trigger step key that <c>connections[].from</c> edges reference.
    /// Only meaningful alongside <see cref="Trigger"/>.
    /// </summary>
    [JsonPropertyName("trigger_key")]
    public string? TriggerKey { get; set; }

    /// <summary>
    /// Update the <c>mailblastr:schedule</c> trigger's schedule
    /// (<c>{at, timezone}</c>). Only valid on automations with that trigger, and
    /// only while the automation is disabled.
    /// </summary>
    [JsonPropertyName("trigger_config")]
    public AutomationTriggerConfig? TriggerConfig { get; set; }

    [JsonPropertyName("connections")]
    public List<AutomationConnection>? Connections { get; set; }
}

/// <summary>
/// Payload for appending a step (POST /automations/:id/steps) or editing one
/// (PATCH /automations/:id/steps/:stepId). The automation must be DISABLED for
/// both. <c>type: "trigger"</c> is rejected — the trigger lives on the
/// automation, not in the step list.
/// </summary>
public class AutomationAddStepOptions
{
    /// <summary>
    /// <c>send_email</c> | <c>wait_for_event</c> | <c>delay</c> | <c>condition</c> |
    /// <c>split</c> | <c>add_to_segment</c> | <c>contact_update</c> |
    /// <c>contact_delete</c>. Optional on PATCH (leaves the type unchanged).
    /// </summary>
    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    [JsonPropertyName("config")]
    public Dictionary<string, object?>? Config { get; set; }

    [JsonPropertyName("key")]
    public string? Key { get; set; }
}

/// <summary>
/// Options for listing automation runs (GET /automations/:id/runs): cursor
/// pagination plus an optional status filter.
/// </summary>
public class AutomationRunListOptions
{
    /// <summary>Page size (1–100; this endpoint defaults to 20).</summary>
    public int? Limit { get; set; }

    /// <summary>Cursor: id of the last item on the previous page.</summary>
    public string? After { get; set; }

    /// <summary>Cursor: id of the first item on the next page.</summary>
    public string? Before { get; set; }

    /// <summary>
    /// Keep only runs in these statuses (<c>running</c>, <c>completed</c>,
    /// <c>failed</c>, <c>skipped</c>). Sent as a comma-separated list; filtering
    /// happens before pagination.
    /// </summary>
    public IEnumerable<string>? Status { get; set; }
}

/// <summary>
/// Prompt for "Create with AI" (POST /automations/:id/ai). The automation must
/// be stopped; without <see cref="Attach"/> it must also have zero steps.
/// </summary>
public class AutomationAiOptions
{
    /// <summary>What the workflow should do. Required; max 2000 characters.</summary>
    [JsonPropertyName("prompt")]
    public string Prompt { get; set; } = null!;

    /// <summary>Templates the plan may send (first 10 kept).</summary>
    [JsonPropertyName("template_ids")]
    public List<string>? TemplateIds { get; set; }

    /// <summary>Event names the plan may wait on (first 10 kept).</summary>
    [JsonPropertyName("events")]
    public List<string>? Events { get; set; }

    /// <summary>
    /// Where to graft the generated steps. Supplying it switches to APPEND mode;
    /// omitting it generates a whole workflow.
    /// </summary>
    [JsonPropertyName("attach")]
    public AutomationAiAttach? Attach { get; set; }
}

/// <summary>Attachment point for an APPEND-mode "Create with AI" request.</summary>
public class AutomationAiAttach
{
    /// <summary>The trigger key or an existing step key to attach after.</summary>
    [JsonPropertyName("from")]
    public string From { get; set; } = null!;

    /// <summary>
    /// <c>default</c> (the default) | <c>condition_met</c> | <c>condition_not_met</c> |
    /// <c>event_received</c> | <c>timeout</c>.
    /// </summary>
    [JsonPropertyName("type")]
    public string? Type { get; set; }

    /// <summary>Insert before this existing step key.</summary>
    [JsonPropertyName("before")]
    public string? Before { get; set; }
}

/// <summary>What a "Create with AI" call added.</summary>
public class AutomationAiInfo
{
    [JsonPropertyName("added_steps")]
    public int AddedSteps { get; set; }

    /// <summary><c>workflow</c> | <c>append</c>.</summary>
    [JsonPropertyName("mode")]
    public string? Mode { get; set; }
}

/// <summary>The automation after a "Create with AI" call, plus what was added.</summary>
public class AutomationAiResult : Automation
{
    [JsonPropertyName("ai")]
    public AutomationAiInfo? Ai { get; set; }
}

/// <summary>Acknowledgement of deleting an automation step: <c>{ id, deleted }</c>.</summary>
public class AutomationStepDeleted
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("deleted")]
    public bool Deleted { get; set; }
}

/// <summary>One executed step within an automation run.</summary>
public class AutomationRunStep
{
    [JsonPropertyName("key")]
    public string Key { get; set; } = null!;

    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    /// <summary><c>completed</c> | <c>failed</c> | <c>skipped</c>.</summary>
    [JsonPropertyName("status")]
    public string Status { get; set; } = null!;

    [JsonPropertyName("started_at")]
    public string? StartedAt { get; set; }

    [JsonPropertyName("completed_at")]
    public string? CompletedAt { get; set; }

    [JsonPropertyName("output")]
    public Dictionary<string, JsonElement>? Output { get; set; }

    [JsonPropertyName("error")]
    public string? Error { get; set; }
}

/// <summary>One enrollment (run) of a contact through an automation.</summary>
public class AutomationRun
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "automation_run";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("contact_id")]
    public string ContactId { get; set; } = null!;

    /// <summary>Email of the contact the run is for; null if that contact was deleted.</summary>
    [JsonPropertyName("contact_email")]
    public string? ContactEmail { get; set; }

    /// <summary><c>running</c> | <c>completed</c> | <c>failed</c> | <c>cancelled</c> | <c>skipped</c>.</summary>
    [JsonPropertyName("status")]
    public string Status { get; set; } = null!;

    [JsonPropertyName("started_at")]
    public string? StartedAt { get; set; }

    [JsonPropertyName("completed_at")]
    public string? CompletedAt { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;

    /// <summary>Present on retrieve (GET /automations/:id/runs/:runId) only.</summary>
    [JsonPropertyName("automation_id")]
    public string? AutomationId { get; set; }

    [JsonPropertyName("steps")]
    public List<AutomationRunStep>? Steps { get; set; }

    [JsonPropertyName("error")]
    public string? Error { get; set; }
}
